pc/registration/icp/evaluator.go

```diff
@@ -102,22 +122,21 @@ func (e *PointToPointEvaluator) Evaluate(base storage.Search, target pc.Vec3Rand
 		pairVecs[0].(pc.Vec3Slice)[i] = pb
 		pairVecs[1].(pc.Vec3Slice)[i] = pt
 
+		w := weightFn(pair.SquaredDistance)
+
+		out.Value += w * pair.SquaredDistance
+		sumWeight += w
+
 		x0, y0, z0 := pt[0], pt[1], pt[2]
 		x1, y1, z1 := pb[0], pb[1], pb[2]
-		out.Value += pair.SquaredDistance
-
-		out.Gradient[0] += 2 * (x0 - x1)
-		out.Gradient[1] += 2 * (y0 - y1)
-		out.Gradient[2] += 2 * (z0 - z1)
-		xd := 2 * (y0*z1 - y1*z0) / (z0*z0 + y0*y0 - 0.5)
-		yd := 2 * (z0*x1 - z1*x0) / (z0*z0 + x0*x0 - 0.5)
-		zd := 2 * (x0*y1 - x1*y0) / (y0*y0 + x0*x0 - 0.5)
-		if !isNaN(xd) && !isNaN(yd) && !isNaN(zd) {
-			out.Gradient[3] += xd
-			out.Gradient[4] += yd
-			out.Gradient[5] += zd
-			num++
-		}
+		out.Gradient[0] += w * 2 * (x0 - x1)
+		out.Gradient[1] += w * 2 * (y0 - y1)
+		out.Gradient[2] += w * 2 * (z0 - z1)
+		out.Gradient[3] += w * 2 * (z0*(y0-y1) - y0*(z0-z1))
+		out.Gradient[4] += w * 2 * (x0*(z0-z1) - z0*(x0-x1))
+		out.Gradient[5] += w * 2 * (y0*(x0-x1) - x0*(y0-y1))
```

Comment:
Isn't the sign flipped here? Shouldn't it be `out.Gradient[3] += w * 2 * (y0*(z0-z1) - z0*(y0-y1))` ? (and same for out.Gradient[4]` and `out.Gradient[5]`)